hivemind/utils/grpc.py

```diff
@@ -1,17 +1,151 @@
 """
 Utilities for running GRPC services: compile protobuf, patch legacy versions, etc
 """
+from __future__ import annotations
+import os
+import threading
+from typing import NamedTuple, Sequence, Tuple, Optional, Union, Any, Dict, TypeVar, Type
 
+import grpc
 import numpy as np
 import torch
 
 from hivemind.proto import runtime_pb2
 from hivemind.proto.runtime_pb2 import CompressionType
+from hivemind.utils.timed_storage import TimedStorage, get_dht_time, DHTExpiration, ValueWithExpiration
+from hivemind.utils.networking import Endpoint
+from hivemind.utils.logging import get_logger
+
+logger = get_logger(__file__)
+
+Stub = TypeVar("Stub")
+
+
+class ChannelInfo(NamedTuple):
+    target: Endpoint
+    aio: bool
+    options: Tuple[Tuple[str, str], ...]
+    credentials: Optional[grpc.ChannelCredentials]
+    compression: Optional[grpc.Compression]
+
+
+class ChannelCache(TimedStorage[ChannelInfo, Tuple[Union[grpc.Channel, grpc.aio.Channel], Dict]]):
+    """
+    A process-wide cache of gRPC channels, supports both normal and aio channels, secure/insecure channels, etc
+    Based on grpcio internal channel cache by Richard Belleville and Lidi Zheng (thanks!)
+    Unlike TimedStorage, ChannelCache actively evicts stale channels even if the cache is not accessed
+    Unlike grpc._simple_stubs.ChannelCache, this implementation supports aio and does not forcibly close active channels
+    """
+    MAXIMUM_CHANNELS = os.environ.get("GRPC_PYTHON_MANAGED_CHANNEL_MAXIMUM", 4096)
+    EVICTION_PERIOD_SECONDS = os.environ.get("GRPC_PYTHON_MANAGED_CHANNEL_EVICTION_SECONDS", 10 * 60)
+    logger.debug(f"Eviction period = {EVICTION_PERIOD_SECONDS}s, max channels = {MAXIMUM_CHANNELS}")
```

Comment:
These constants should be module-level

Reply:
(fixed, thanks!)

Reply:
upd - it complicates testing, added to #98